### ISSUE TITLE
SS_readpar_3.30() fails for model with semi-parametric selectivity

### ISSUE THREAD

**Issue by iantaylor-NOAA** [r4ss/r4ss#681]
In the `test_readwrite_par` branch I added SS_readpar_3.30() to the general test of read/write functions (see https://github.com/r4ss/r4ss/commit/c58ed939056d4e3f06daa8dc3113bbb48ba11499) and see that it's failing (see https://github.com/r4ss/r4ss/runs/6866303767?check_suite_focus=true#step:13:273). 

It's not clear from the github action which of the test models is the source of the failure, but running the function on all those models locally, I see that it's just "Hake_2019_semi-parametric_selex" (files at  https://github.com/nmfs-stock-synthesis/test-models/tree/main/models/Hake_2019_semi-parametric_selex) which fails and the issue is that the semi-parametric (a.k.a. 2D-AR1) selectivity setup creates a `parm_dev` vector in ss.par but unlike other cases there are no parameter lines with a non-zero value in column 9 to trigger filling in the `dev_parm_start` and `dev_parm_end` values that the function is expecting.

@nathanvaughan-NOAA, do you want to take this on as the expert on `SS_readpar_3.30()`, or would you rather that I work on the fix? Either way is fine with me.

Also, @k-doering-NOAA, do you have any tips on figuring out which of the test models causes a failure in the testthat output?

**Comment by nathanvaughan-NOAA** [comment]
Thanks for the detailed investigation @iantaylor-NOAA I can take a crack at it.

**Comment by k-doering-NOAA** [comment]
@iantaylor-NOAA, thanks for pointing this out! I wonder if we could print the name of the model before starting a new test, and that would show up in the console output in github actions, maybe? I suspect I have done as you have in the past, testing locally.

**Comment by nathanvaughan-NOAA** [comment]
@iantaylor-NOAA  and @k-doering-NOAA  I made the correction to the par function but it kept failing. After looking into the tests locally it seems that the issue is in the test-readwrite.R test. I there the SS_read() and SS_write() functions do not read or write the par file. However the final check lapply(files, function(x) expect_true(file.exists(x))) expects the par file to exist. I will add the par file into these functions and see if that clears it up. I'm not sure why this test wasn't failing for every model though??

**Comment by iantaylor-NOAA** [comment]
Thanks @nathanvaughan-NOAA for fixing this so quickly in PR #684.

When modifying the test to add the read/write of the par file I didn't think about the missing link to `SS_read()` and `SS_write()`. 
Here's a question for you and other r4ss users.

I think it's useful to have the par included in the those two wrapper read/write functions along with the other input files. However, it could also cause a problem in one of the workflows that I imagine:
1. run `SS_read()` to bring input files into R workspace
2. modify the input files somehow
3. run `SS_write()` to create modified input text files
4. run Stock Synthesis, perhaps via `run_SS_models()`
5. repeat

I'm guessing that in the majority of cases (or at least in my own previous assessment work with the functions that preceded `SS_read()` and `SS_write()`), that iterative process won't involve the .par file (though I typically use it at a later stage when modifying forecasts, for which the new functionality will be helpful). As currently configured, the `SS_write()` step could break if you modify the control file in a way that makes it out of sync with the .par file (any change that adds or subtracts a parameter). Likewise if the user modifies the control file manually and doesn't make a matching change to the .par file, then the `SS_read()` step could break.

Here are some options for solutions. Anyone please chime in with input on what you think makes the most sense
1. expect the user to delete the ss.par file if they don't want to use it since the revised function wraps the read of it within `file.exists()`
2. wrap the read/write of the .par into a `try()` so that the function doesn't fail if there's a mismatch between the .par and other files.
3. give the user a control on whether to read/write the .par and make the default FALSE so you only deal with it when you want to.
4. read or write the .par file if `start[["init_values_src"]] == 1` and ignore otherwise (this could be combined with the `try()` in option 2).

**Comment by nathanvaughan-NOAA** [comment]
Good catch @iantaylor-NOAA , I hadn't really thought about the use case for the SS_read and SS_write functions. I think the try() wrapper sounds like the best solution in this case for now. It may also be helpful to have it set start[["init_values_src"]] == 0 if it can't write the par file and give the user a warning. I can push that change while the pull request is still open if we all agree.

**Comment by kellijohnson-NOAA** [comment]
My thoughts
1. I do not think that this is a viable assumption or something that I would like to require myself to do in my workflow.
2. I think that it is potentially fine to have the par file in the list even if it is not in sync with the control file.
3. I would prefer to not add more input arguments.
4. I think if we do this option that the "init_values_src" should only be consulted for the SS_write to determine if the par file should be written.

**Comment by iantaylor-NOAA** [comment]
Thanks for the input everyone. Here's a refined list. This discussion has probably already taken up too much time and whatever we end up implementing will be fine unless we decide it isn't and change it in the future. Feel free to chime in with more thoughts. If there are none, I can make a decision tomorrow.

1. ~**delete the .par manually**~ I agree with @kellijohnson-NOAA that this is a bad option. I'm not sure why I even included it on the list.
2. ~**reading the .par file within try()**~ `SS_readpar_3.30()` requires first reading the control and data files to get the model dimensions and number of parameters, so if those files get out of sync with the .par, then the read will crash rather than just exist as a mismatched list element. It could still succeed and be mismatched if you added one parameter and subtracted another but potentially crash the model if it ran with the resulting .par file. On further thought, I don't like the idea of having to constantly ignore a message from try() about a failed .par file being read, so I'm going to vote against this option.
3. **add additional argument like par = TRUE/FALSE** Fair point @kellijohnson-NOAA to minimize the number of arguments, although some other functions have such a bloated list of arguments, I'm not sure being super strict in this case will make much difference for this package which is so messy in so many other places. 
Having said that, I recently considered adding an additional argument to `SS_read()` that would allow you to read `data_expval.ss` or `data_boot_*.ss` instead of the original input file or `data_echo.ss_new` but then I realized that my reasons for wanting that other file were not very common and it was easy enough to just add an extra line of code to get it.
4. **only read .par if being used by the model** this would match the status-quo setup for the wtatage file which only gets read if the control file indicates that the model uses that file. If you change your model configuration from one that doesn't require a wtatage file to one that does, you would have to manually add a step of dealing with the wtatage file. Likewise, if you change the model configuration to require a .par file, you could either add the step of reading it into R or trust that the existing .par file contains what you need from the previous model run.
5. **(new option) make users call `SS_readpar_3.30()` and `SS_writepar_3.30()` directly if they want to work with the .par file instead of including it in `SS_read()` and `SS_write()`.** This would just return to the old setup before the testing issue brought all this up. If `ss.par` is considered an input file then this goes against the idea of these wrapper functions dealing with all the input files, but it would keep the workflow cleaner than the options above for all the cases where you aren't using it.

**Comment by nathanvaughan-NOAA** [comment]
I would lean towards considering ss.par an input file (even if it is one that most people don't modify). Regarding the various options, my standing would be
1) Agree this is a bad option let's not do it.
2) I suggest leaving the try loop but adding silent=TRUE so that no messages are returned. Obviously, if the par list doesn't exist the user knows it failed.
3) I would suggest if we want the optional argument to set a default read_par=TRUE. Combined with the try loop this has no downside if the par doesn't exist or is mismatched and if the par is read but not needed users can always ignore it. This way there may be an extra argument option but it is one people don't even need to know is there. Does anyone have a scenario where there is any real downside to reading in the par file?
4) I lean against this option as it is reasonable that someone may read in their options with the intention of changing the read from par to 1 in the code. It seems counterproductive to force the user to modify the control file by hand in a function designed to facilitate automation.
5) As stated above I don't think this is a good idea. Requiring multiple function calls to avoid adding an additional argument to this function seems counterproductive.

In summary, I think a silent try wrapper on each of the par read/write calls is the best option. We can also add the read_par argument if desired. I have all of these options coded already and am happy to push them as soon as we make a decision on which ones to include. Standing by for an executive decision @iantaylor-NOAA  :)

**Comment by kellijohnson-NOAA** [comment]
@nathanvaughan-NOAA what are the differences between `try()` and `tryCatch()`? I tend to use the latter where I can put in any message/option I want if there is an error.

**Comment by iantaylor-NOAA** [comment]
I was the one who suggested `try()`, when I meant to suggest `tryCatch()`. I don't actually know the difference.

**Comment by nathanvaughan-NOAA** [comment]
tryCatch() is basically just try() with the option to do something after an error is caught. Given that we don't want to take any actions on the try() it should suffice. I had also set up a tryCatch() option that modified the starter to not read from par if writing the par file failed but I think that is more complexity than we want.

**Comment by nathanvaughan-NOAA** [comment]
@iantaylor-NOAA I pushed the new code with try() wrappers and your silent=!verbose idea. It has passed the tests on my fork, so ready for you to review when you have a chance. Enjoy your vacation!

**Comment by k-doering-NOAA** [comment]
@iantaylor-NOAA , can this one be closed?

**Comment by iantaylor-NOAA** [comment]
Oops, yes, resolved thanks to @nathanvaughan-NOAA with PRs #684 and #689.
